# Identifying Regulatory Degeneration:

Here are the commands used to identify the regulatory degeneration of _P. reticulata_ following the attainment of scRNA-seq data from [Darolti & Mank (2023)](https://academic.oup.com/evlett/article/7/3/148/7119960). 
----------------------------------------------------------

## Prepping VCF files from the sorted BAM files

We used the possorted_genome_bam.bam files from the scRNA seq data available from Darolti & Mank (2023). We used  **[cellSNP-lite](https://github.com/single-cell-genetics/cellSNP)** to generate our VCF files. We have barcode info for each library, so we did one BAM library at a time with the barcode information. cellSNP command is run as: cellsnp-lite -s $BAM -b $BARCODE -O $OUT_DIR -p 10 --minMAF 0.1 --minCOUNT 100 --gzip --chrom
Here is an example command:

    cellsnp-lite -s sample6_possorted_genome_bam.bam -b sample6_barcodes.tsv.gz -O ~/scRNA_cellsnp_files/Lib6 --gzip -f Poecilia_reticulata.Guppy_female_1.0_MT.dna.toplevel.fa --minMAF 0.01 --minCOUNT 10 -p 16 --maxDEPTH 100000 --chrom LG1,LG2,LG3,LG4,LG5,LG6,LG7,LG8,LG9,LG10,LG11,LG12,LG13,LG14,LG15,LG16,LG17,LG18,LG19,LG20,LG21,LG22,LG23

Following VCF generation, we re-formatted the VCF files to fit the format for **[scAlleleCount](https://github.com/barkasn/scAlleleCount)**. Example command below:

        gunzip Lib6/cellSNP.base.vcf.gz
        awk '{if ($1 ~/^LG/) print $1, $2, $4, $5}' Lib6/cellSNP.base.vcf > format_VCFs/sample6_snps_scAC.txt

        
        
We then used **[SAMtools](https://github.com/samtools/samtools)** to prepare the barcode file from the BAM file to run for scAlleleCount. Example command:

        samtools view sample6_possorted_genome_bam.bam | grep -o -P 'CB:Z:.{18}' > sample6_barcodes.txt
        sed -i 's/CB:Z://' sample6_barcodes.txt

        
We ran a modified scAlleleCount python script - see [modified_scAlleleCount_script_example.py](https://github.com/ljmfong/Pret_Y_degeneration/blob/main/modified_scAlleleCount_script_example.py). The outputted matricies were then processed in R - see [R_commands_to_prep_matrices.R](https://github.com/ljmfong/Pret_Y_degeneration/blob/main/R_commands_to_prep_matrices.R).

After attainmaint of the matrices, we processed each tissue separately - see (prepping_scRNA_scripts)[].

