# Predicting Protein Folding Changes:

Here are the commands used to predict the changes in protein folding/conformation using **[FoldX](https://foldxsuite.crg.eu/documentation)**.
----------------------------------------------------------

## Obtaining the protein models

We downloaded PDB files from **[AlphaFold](https://alphafold.ebi.ac.uk/)** for genes that were predicted to contain deleterious SNPs. If AlphaFold did not have a protein model, we used **[SWISS-MODEL](https://swissmodel.expasy.org/)** to predict the protein structure from protein sequences obtained from **[UniProt](https://www.uniprot.org/)**.

## Running FoldX

We used the amino acid changes predicted by snpEff to build the variant model for proteins. See [Example files]() for example mutant file used for FoldX. Here is an example command:

    .\foldx_20251231.exe -c BuildModel --pdb=spp1-model_v6.pdb --mutant-file=mutant_file_list\individual_list_spp1.txt --output-dir=build_model_output
    .\foldx_20251231.exe -c Stability --pdb=spp1-model_v6.pdb
    .\foldx_20251231.exe -c Stability --pdb=spp1-model_v6_1.pdb

Wherein the "model_v6.pdb" is the wildtype file and the "model_v6_1.pdb" is the variant model built by FoldX. To calculate the change in Stability, we used the Total Energy outputted by the Stability command.
